src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java

```diff
@@ -256,70 +296,120 @@ public int getConnectTimeout() {
     }
 
     @Override
+    public int getMaxPayloadLength() {
+        return maxPayloadLength;
+    }
+
+    @Override
     public HttpRedirectPolicy getRedirectPolicy() {
         return connection.getRedirectPolicy();
     }
 
     @Override
     public WsInputStream getInputStream() throws IOException {
-        if (inputStream == null) {
+        if (inputStream != null) {
+            return inputStream;
+        }
+
+        try {
+            stateLock.lock();
             ensureConnected();
             inputStream = new WsInputStream(this);
+            return inputStream;
+        }
+        finally {
+            stateLock.unlock();
         }
-
-        return inputStream;
     }
 
-    @Override
     public WsMessageReader getMessageReader() throws IOException {
-        if (messageReader == null) {
+        if (messageReader != null) {
+            return messageReader;
+        }
+
+        try {
+            stateLock.lock();
             ensureConnected();
             messageReader = new WsMessageReader(this);
```

Comment:
Do we need another check for non-null inside the lock here, just in case another thread racing us has already created the message reader and exited the lock.